components/object_store/src/mem_cache.rs

```diff
@@ -0,0 +1,322 @@
+        // don't care error now.
+        _ = guard.insert(key, CachedBytes(value));
+    }
+}
+
+#[derive(Debug)]
+struct MemCache {
+    /// Max memory this store can use
+    mem_cap: usize,
+    partitions: Vec<Arc<Partition>>,
+    partition_mask: usize,
+}
+
+impl MemCache {
+    fn new(partition_bits: usize, mem_cap: usize) -> Self {
+        let partition_num = 1 << partition_bits;
+        let cap_per_part = mem_cap / partition_num;
+        let partitions = (0..partition_num)
+            .map(|_| Arc::new(Partition::new(cap_per_part)))
+            .collect::<Vec<_>>();
+
+        Self {
+            mem_cap,
+            partitions,
+            partition_mask: partition_num - 1,
+        }
+    }
+
+    fn locate_partition(&self, key: &String) -> Arc<Partition> {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        self.partitions[hasher.finish() as usize & self.partition_mask].clone()
+    }
+
+    async fn get(&self, key: &String) -> Option<Bytes> {
+        let partition = self.locate_partition(key);
+        partition.get(key).await
+    }
+
+    async fn insert(&self, key: String, value: Bytes) {
+        let partition = self.locate_partition(&key);
+        partition.insert(key, value).await;
+    }
+}
+
+impl Display for MemCache {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MemCache")
+            .field("mem_cap", &self.mem_cap)
+            .field("mask", &self.partition_mask)
+            .field("partitons", &self.partitions)
+            .finish()
+    }
+}
+
+#[derive(Debug)]
+pub struct CachedStore {
+    cache: MemCache,
+    underlying_store: Arc<dyn ObjectStore>,
+}
+
+impl CachedStore {
+    pub fn new(
+        partition_bits: usize,
+        mem_cap: usize,
+        underlying_store: Arc<dyn ObjectStore>,
+    ) -> Self {
+        Self {
+            cache: MemCache::new(partition_bits, mem_cap),
+            underlying_store,
+        }
+    }
+
+    fn cache_key(location: &Path, range: &Range<usize>) -> String {
+        format!("{}-{}-{}", location, range.start, range.end)
+    }
+}
+
+impl Display for CachedStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.cache.fmt(f)
+    }
+}
+
+#[async_trait]
+impl ObjectStore for CachedStore {
+    async fn put(&self, location: &Path, bytes: Bytes) -> Result<()> {
+        self.underlying_store.put(location, bytes).await
+    }
+
+    async fn put_multipart(
+        &self,
+        location: &Path,
+    ) -> Result<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)> {
+        self.underlying_store.put_multipart(location).await
+    }
+
+    async fn abort_multipart(&self, location: &Path, multipart_id: &MultipartId) -> Result<()> {
+        self.underlying_store
+            .abort_multipart(location, multipart_id)
+            .await
+    }
+
+    async fn get(&self, location: &Path) -> Result<GetResult> {
```

Comment:
How about adding comments to explain why we don't cache it while get all.

Reply:
fixed